core/types/receipt.go

```diff
@@ -198,32 +204,82 @@ func (r *ReceiptForStorage) DecodeRLP(s *rlp.Stream) error {
 	if err != nil {
 		return err
 	}
-	var dec receiptStorageRLP
-	if err := rlp.DecodeBytes(blob, &dec); err != nil {
-		var sdec LegacyReceiptStorageRLP
-		if err := rlp.DecodeBytes(blob, &sdec); err != nil {
-			return err
-		}
-		dec.PostStateOrStatus = common.CopyBytes(sdec.PostStateOrStatus)
-		dec.CumulativeGasUsed = sdec.CumulativeGasUsed
-		dec.TxHash = sdec.TxHash
-		dec.ContractAddress = sdec.ContractAddress
-		dec.Logs = sdec.Logs
-		dec.GasUsed = sdec.GasUsed
-	}
-	if err := (*Receipt)(r).setStatus(dec.PostStateOrStatus); err != nil {
+
+	if err := decodeStoredReceiptRLP(r, blob); err == nil {
+		return nil
+	}
+
+	if err := decodeV4StoredReceiptRLP(r, blob); err == nil {
+		return nil
+	}
+
+	return decodeLegacyStoredReceiptRLP(r, blob)
+}
+
+func decodeStoredReceiptRLP(r *ReceiptForStorage, blob []byte) error {
+	var stored storedReceiptRLP
+	if err := rlp.DecodeBytes(blob, &stored); err != nil {
+		return err
+	}
+
+	if err := (*Receipt)(r).setStatus(stored.PostStateOrStatus); err != nil {
 		return err
 	}
-	// Assign the consensus fields
-	r.CumulativeGasUsed = dec.CumulativeGasUsed
-	r.Logs = make([]*Log, len(dec.Logs))
-	for i, log := range dec.Logs {
+
+	r.CumulativeGasUsed = stored.CumulativeGasUsed
+	r.ContractAddress = stored.ContractAddress
+	r.Logs = make([]*Log, len(stored.Logs))
+	for i, log := range stored.Logs {
 		r.Logs[i] = (*Log)(log)
 	}
+	r.Bloom = CreateBloom(Receipts{(*Receipt)(r)})
 
+	return nil
+}
+
+func decodeV4StoredReceiptRLP(r *ReceiptForStorage, blob []byte) error {
+	var stored v4StoredReceiptRLP
+	if err := rlp.DecodeBytes(blob, &stored); err != nil {
+		return err
+	}
+
+	if err := (*Receipt)(r).setStatus(stored.PostStateOrStatus); err != nil {
+		return err
+	}
+
+	r.CumulativeGasUsed = stored.CumulativeGasUsed
```

Comment:
I realize there is some code duplication amongst these methods, but I felt like it would good for everything to be self contained.